### ISSUE TITLE
Add support for SAM3

### ISSUE THREAD

**Issue by LinkZY3471** [heyoeyo/muggled_sam#23]
Amazing work! Is it possible to add support for SAM3? :)

**Comment by heyoeyo** [comment]
Thanks! Actually I hadn't realized SAM3 was released, so thanks for the heads up.
Just looking at it briefly, the image segmentation might not be too hard to support, but the video segmentation seems to have a lot of changes so that might take a while. But ya, I'll probably have a look at it over the next few weeks.

**Comment by claydugo** [comment]
Wanted to report back that your reading was correct and your psuedocode was very helpful.

I've had the best results with keeping the encodings and then using a text prompt + exemplar encoding. 
It does unfortunately seem to matter if the text prompt is relevant to the thing you want to segment.
I was optimistic you could just lean on the the initial prompt and encodings and the text wouldn't matter.
For the above example `"cells"` seems to do pretty well.


To get around this I did try to just add a box the size of the whole image and it had okay results. Worse than text for sure.

Would love to hear if you play around with this!

**Comment by LinkZY3471** [comment]
I'm using sam3 source code (PVS, box prompts) for remote sensing images (to segment farmland parcels), and it can get better results with less noise compared to sam1. So I think in some specific domains, sam3 does improve mask quality. However, I found sam3 is indeed much slower than sam1.

<img width="1902" height="880" alt="Image" src="https://github.com/user-attachments/assets/4fa2bd7c-b8eb-4ed1-af09-ce27cce15bd3" />

<img width="1826" height="906" alt="Image" src="https://github.com/user-attachments/assets/60b148bc-c1e7-48c7-ba20-2977b0e7b030" />

**Comment by heyoeyo** [comment]
> I wonder why there would be such a discrepancy between our results.

I'd be interested to know what could be causing this as well!
For reference, I'm mostly basing my results off the [speed benchmark](https://github.com/heyoeyo/muggled_sam/tree/main/simple_examples#speed-benchmarking)/[video segmentation](https://github.com/heyoeyo/muggled_sam/tree/main/simple_examples#video-segmentation) example scripts. I've yet to do a careful comparison of mugsam vs the original samv3, it's possible the original repo runs faster and that might explain the discrepency?

**Comment by heyoeyo** [comment]
Hey everyone, the repo has been updated with full support for the detection features of SAMv3. The usage is similar to the existing masking, basically a 3 step sequence: 'encode image', 'encode exemplars (prompts)' and then 'generate detections':

```python
# Load model
model_config_dict, base_model = make_sam_from_state_dict("path/to/sam3.pt")
detmodel = base_model.make_detector_model() # ***NEW***

# Run detection
encoded_imgs, _, _ = detmodel.encode_detection_image(image)
encoded_exemplars = detmodel.encode_exemplars(encoded_imgs, text_prompt, ...)
mask_preds, box_preds, det_scores, pres_score = detmodel.generate_detections(
  encoded_imgs, encoded_exemplars
)
```

The only unusual part is the second line, where detector instance gets created. I'm not sure if I'll keep it this way, but it helps avoid having too many similar looking (but incompatible) functions on the same instance. For a more complete version, see the new [object detection](https://github.com/heyoeyo/muggled_sam/blob/main/simple_examples/object_detection.py) example.

There's also a very basic interactive demo (run_detections.py):

https://github.com/user-attachments/assets/f5bbd59a-9a3a-4eaf-8f67-7158246735f5

The idea is to input 'prompts' (exemplars) on the left and detections/masks are then shown on the right. Text input is somewhat awkwardly handled using the terminal, since the OpenCV UI isn't great for this (the video above has an example of this at the end). There's also a bit of a strange default text prompt of 'visual', which comes from the original code base. Point/box detections tend to perform worse without this, though you can check this by toggling the `Use Text` button on the UI.
Real-time visualization of results is turned off by default (it's *much* heavier than the prior masking functionality), but should be usable if you're running on a GPU.

On the topic of weird defaults, when encoding point/box prompts, the original model includes the coordinates themselves as part of the 'exemplar encoding'. This can be disabled with the `Include Coords` toggle button, though this often degrades results. However, it may help with cross-image detections (@claydugo might be relevant for what you were describing?). The demo supports cross-image detections by providing a `-r path/to/reference_image` flag when launching the script.

The detection UI doesn't support video tracking and I'm guessing it would be a month or more before I'd get to making something for that. However, I'll probably try to add a 'simple example' script for how to do it within the next few days. For anyone wanting to try it sooner, I think it should just be a combination of the [video segmentation from mask](https://github.com/heyoeyo/muggled_sam/tree/main/simple_examples#video-segmentation-from-mask) (using the v3 detection masks) and [multi-object video segmentation](https://github.com/heyoeyo/muggled_sam/tree/main/simple_examples#video-segmentation-multi-object) examples. Interestingly, it should be possible to use the v3 model for detection masks and then a (much faster) v2 model for tracking.

**Comment by claydugo** [comment]
Thanks so much for this! 

I have found your tools to be so useful to look into these black boxes.

I tested the cross-image `-r` argument and it seems to work great. 

My big learning from playing with this tonight was that the cross-image box prompts perform _substantially_ worse than their point counterparts.

<details>
  <summary>Box</summary>


<img width="1490" height="1346" alt="Image" src="https://github.com/user-attachments/assets/0ed356d3-233e-4b3d-aa01-67c775e280df" />

</details>

<details>
  <summary>Point</summary>

<img width="1490" height="1346" alt="Image" src="https://github.com/user-attachments/assets/fff2de2b-dfb5-4671-8c51-5d2852191e15" />

</details>

Which explains some of the headaches I had with my initial exploration. 

Your UI is so great for visualizing these workflows. Thanks again!

**Comment by heyoeyo** [comment]
Thanks for the kind words!

That's interesting that the boxes behave so differently, I hadn't noticed that while I was playing around with it. I'm still trying to get a sense of how all the different pieces (box+point+text+coords) affect one another...

**Comment by skylning** [comment]
@LinkZY3471 Hi man,can you tell me what the resolution of the remote sensing images and your prompt?i use the official  SAM3 to inference 0.3m remote sensing images,and the result is bad

**Comment by heyoeyo** [comment]
Hi all, just closing this issue for now since the SAMv3 support is more or less completed. Feel free to open other issues if needed. There are also now [distillation scripts](https://github.com/heyoeyo/muggled_sam/tree/main/training) that can be used to make smaller/faster versions of SAMv3, in case that's of interest to anyone.
